[#6186][feat] Introduce QKNormRoPEAttention module

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Qwen3 models now use a redesigned self-attention that integrates Q/K normalization with rotary positional embeddings for improved stability and output quality, including MoE variants.

* **Performance Improvements**
  * Faster inference via a fused attention execution path and parallelized normalization, reducing latency and compute overhead.

* **Compatibility**
  * No user-facing API changes; existing Qwen3 workflows remain supported.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up
-->

## Description

- Added QKNormRoPEAttention class to apply QK normalization and RoPE.
- Replaced Qwen3Attention with QKNormRoPEAttention in Qwen3DecoderLayer.
- Removed unused imports and code related to the previous attention implementation.

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
[feat] Introduce QKNormRoPEAttention class for enhanced attention mechanism in Qwen3 model

- Added QKNormRoPEAttention class to apply QK normalization and RoPE.
- Replaced Qwen3Attention with QKNormRoPEAttention in Qwen3DecoderLayer and Qwen3MoEDecoderLayer.
- Removed unused imports and code related to the previous attention implementation.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #14981](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14981/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #14981](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/14981/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #11311](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11311/) completed with status: 'FAILURE'

Funatiq committed:
[feat] Refactor QKNormRoPEAttention initialization in Qwen3 model layers

- Updated QKNormRoPEAttention constructor to accept parameters directly instead of a model_config object.
- Modified Qwen3DecoderLayer and Qwen3MoEDecoderLayer to pass specific configuration values to QKNormRoPEAttention.
- Enhanced clarity and flexibility in attention layer configuration.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15115](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15115/) [ run ] triggered by Bot

Funatiq committed:
[feat] Add SPDX license headers to QKNormAttention module

- Included SPDX license headers for compliance and clarity.
- Ensured proper licensing information is present in the QKNormAttention implementation.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #15137](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15137/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15117](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15117/) [ run ] completed with state `ABORTED`
[/LLM/main/L0_MergeRequest_PR pipeline #11413](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11413/) completed with status: 'FAILURE'

yuxianq commented on tensorrt_llm/_torch/modules/qk_norm_attention.py:
```diff
@@ -0,0 +1,135 @@
+from ..modules.rms_norm import RMSNorm
+
+
+class QKNormRoPEAttention(Attention):
```
Can you also use `QKNormRoPEAttention` for Gemma3 and Exaone4 to validate whether it is general enough?

yuxianq commented on tensorrt_llm/_torch/modules/qk_norm_attention.py:
```diff
@@ -0,0 +1,135 @@
+            )
+        else:
+            pos_embd_params = PositionalEmbeddingParams(
+                type=PositionEmbeddingType.rope_gpt_neox,
```
We can move the `pos_embd_params` initialization out of `QKNormRoPEAttention` to make it general since different model uses different default `PositionEmbeddingType` for `pos_embd_params`.

yuxianq commented on tensorrt_llm/_torch/models/modeling_qwen3.py:
```diff
@@ -4,130 +4,41 @@
-            )
-        else:
-            pos_embd_params = PositionalEmbeddingParams(
-                type=PositionEmbeddingType.rope_gpt_neox,
```
We can still keep `Qwen3Attention(QKNormRoPEAttention)` to hold the `pos_embd_params` initialization part like other models.

Funatiq replied:
Done.

Funatiq committed:
[feat] Restore Qwen3Attention class as a subclass of QKNormRoPEAttention

- Enhanced QKNormRoPEAttention to accept positional embedding parameters.
- Inherit QKNormRoPEAttention in Qwen3Attention, specialize it for Qwen3 model.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Refactor Gemma3Attention to inherit from QKNormRoPEAttention

- Updated Gemma3Attention class to extend QKNormRoPEAttention, enhancing its functionality.
- Removed redundant QK normalization methods and integrated them into the new parent class.
- Adjusted initialization parameters to align with the new attention mechanism.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Refactor Exaone4Attention to inherit from QKNormRoPEAttention

- Updated Exaone4Attention class to extend QKNormRoPEAttention, streamlining the attention mechanism.
- Removed redundant QK normalization methods and integrated functionality into the parent class.
- Adjusted initialization parameters for improved clarity and flexibility in attention layer configuration.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #15800](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15800/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #15814](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15814/) [ run ] triggered by Bot

brb-nv commented on tensorrt_llm/_torch/modules/qk_norm_attention.py:
```diff
@@ -0,0 +1,242 @@
+        attention_factor: float, the post-processing scaling factor applied to the computed cos/sin
+    """
+
+    # The config does not contain rope_scaling, which means the model is not using yarn
```
Even if there is scaling, it's not necessary that the model uses yarn, I think? Pls correct me if I'm wrong.

Funatiq replied:
@byshiue can you comment please? It would be nice if we can generalize the yarn path.

brb-nv reviewed (approved):
Few questions for my own understanding.

tensorrt-cicd commented:
[PR_Github #15814](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/15814/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #11887](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/11887/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/11887/test-results/rerun_report.html) for details.

Funatiq committed:
[fix] Update yarn parameter computation in QK normalization attention

- Enhanced the logic to check for yarn usage by verifying both rope_scaling and rope_type in the configuration.
- Improved code clarity by updating comments to reflect the new checks.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #16878](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16878/) [ run ] triggered by Bot

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16894](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16894/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #16894](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16894/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #12691](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/12691/) completed with status: 'FAILURE'

yuxianq commented on tensorrt_llm/_torch/models/modeling_qwen3.py:
```diff
@@ -156,86 +48,28 @@ def __init__(
         self,
         model_config: ModelConfig[Qwen3Config],
         layer_idx: int,
-    ) -> Tuple[torch.Tensor, torch.Tensor]:
```
Why remove ` -> Tuple[torch.Tensor, torch.Tensor]`?

Funatiq replied:
Because the init function doesn't return anything.

yuxianq replied:
Sure.

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #16974](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/16974/) [ run ] triggered by Bot
